### PR TITLE
Use more smart pointers in Source/WebKit

### PR DESCRIPTION
**Pull request by cdumez**
#### 7763e295bcf62a0fd798c716a36684bcab236162
<pre>
Use more smart pointers in Source/WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=261518">https://bugs.webkit.org/show_bug.cgi?id=261518</a>

Reviewed by Darin Adler.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setFocusedElement):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setFocusedElementIfNeeded):
* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::confirmMarkedText):
* Source/WebCore/html/RadioInputType.cpp:
(WebCore::RadioInputType::handleKeydownEvent):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::setCapturingMouseEventsElement):
(WebCore::EventHandler::dispatchMouseEvent):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::setFocusedElement):
* Source/WebCore/page/FocusController.h:
(WebCore::FocusController::setFocusedElement):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragmentInternal):
* Source/WebCore/rendering/RenderEmbeddedObject.cpp:
(WebCore::RenderEmbeddedObject::handleUnavailablePluginIndicatorEvent):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::getOrCreate):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCProvider.cpp:
(WebKit::LibWebRTCProvider::createSocketFactory):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCProvider.h:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCResolver.cpp:
(WebKit::LibWebRTCResolver::sendOnMainThread):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocket.cpp:
(WebKit::LibWebRTCSocket::Close):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocketFactory.cpp:
(WebKit::LibWebRTCSocketFactory::setConnection):
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::sendMessage):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(-[WKPDFPluginAccessibilityObject parent]):
(-[WKPDFPluginAccessibilityObject accessibilityFocusedUIElement]):
(-[WKPDFPluginAccessibilityObject accessibilityAssociatedControlForAnnotation:]):
(WebKit::PDFPlugin::PDFPlugin):
(WebKit::PDFPlugin::getResourceBytesAtPosition):
(WebKit::PDFPlugin::jsPDFDocPrint):
(WebKit::PDFPlugin::handleContextMenuEvent):
(WebKit::PDFPlugin::isFullFramePlugin const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.h:
(WebKit::PDFPluginAnnotation::PDFPluginAnnotation):
(WebKit::PDFPluginAnnotation::parent const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.mm:
(WebKit::PDFPluginAnnotation::attach):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginChoiceAnnotation.mm:
(WebKit::PDFPluginChoiceAnnotation::updateGeometry):
(WebKit::PDFPluginChoiceAnnotation::createAnnotationElement):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::handleEvent):
(WebKit::PluginView::clipRectInWindowCoordinates const):
(WebKit::PluginView::focusPluginElement):
* Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::createSource):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::focusServiceWorkerClient):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::focusTextInputContextAndPlaceCaret):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView clearFocus]):
(-[WebHTMLView becomeFirstResponder]):

Canonical link: <a href="https://commits.webkit.org/267992@main">https://commits.webkit.org/267992@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/cbf55e15fb24d9bb4f133aa8bcdebdbd6be2d249

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20122 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17111 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18477 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21915 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18770 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19046 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18504 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18716 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15920 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21005 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15939 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16673 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23170 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16958 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16844 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21059 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17411 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14769 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16507 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4361 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20870 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17254 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/editing/ios/EditorIOS.mm [hunk 1]
```diff
@@ -307,11 +307,11 @@ static void getImage(Element& imageElement, RefPtr<Image>& image, CachedImage*&
 {
     // FIXME: This is a hacky workaround for the keyboard calling this method too late -
     // after the selection and focus have already changed. See <rdar://problem/5975559>.
-    Element* focused = document().focusedElement();
-    Node* composition = compositionNode();
-    if (composition && focused && focused != composition && !composition->isDescendantOrShadowDescendantOf(focused)) {
```
Comment (cdumez) [inline Source/WebCore/editing/ios/EditorIOS.mm:312]: `focused != composition` is true unconditionally since `composition` is a `Text*` and `focused` is an `Element*` and those 2 types are not related.

File: Source/WebCore/editing/FrameSelection.cpp [hunk 2]
```diff
@@ -2372,19 +2372,19 @@ void FrameSelection::setFocusedElementIfNeeded()
...
-    if (Element* target = m_selection.rootEditableElement()) {
+    if (RefPtr target = m_selection.rootEditableElement()) {
         // Walk up the DOM tree to search for an element to focus.
         while (target) {
             // We don't want to set focus on a subframe when selecting in a parent frame,
             // so add the !isFrameElement check here. There's probably a better way to make this
             // work in the long term, but this is the safest fix at this time.
-            if (target->isMouseFocusable() && !isFrameElement(target)) {
-                CheckedRef(m_document->page()->focusController())->setFocusedElement(target, *m_document->frame());
+            if (target->isMouseFocusable() && !isFrameElement(target.get())) {
```
Comment (darinadler) [inline Source/WebCore/editing/FrameSelection.cpp]: Seems like `isFrameElement` should take a `Node&`.

File: Source/WebKit/WebProcess/Network/webrtc/LibWebRTCProvider.h [hunk 3]
```diff
@@ -33,6 +33,9 @@
 
 #if USE(LIBWEBRTC)
 
+#include "WebPage.h"
```
Comment (darinadler) [inline Source/WebKit/WebProcess/Network/webrtc/LibWebRTCProvider.h]: Is the use of smart pointers going to lead to a lot more includes like this? We should try to avoid the large compile time cost increase that could cause across the project if we can. Maybe forward declarations will work as long as we move appropriate functions out of the header?

**Review by darinadler — APPROVED** [review]
I am surprised by how often we have to write `copyRef`.

Without a static analyzer output I have a lot of difficulty knowing whether this is too much, too little, or just right amount of smart pointers. It's unclear to me when we change the type of a function to take a `RefPtr&&`, for example, how we choose that over `RefPtr` or a raw pointer.

**Comment by cdumez** [comment]
> I am surprised by how often we have to write `copyRef`.

I'll double check, maybe some of them are optional.

> Without a static analyzer output I have a lot of difficulty knowing whether this is too much, too little, or just right amount of smart pointers.

This patch is based on the static analyzer output.

> It's unclear to me when we change the type of a function to take a `RefPtr&&`, for example, how we choose that over `RefPtr` or a raw pointer.

We normally don't need to change the parameter types for this effort. What matters is that the caller is holding a local Ref/RefPtr to the object it is calling a member function on and to the parameters it is passing to the function.

I did however change some parameter types to RefPtr&& because they seemed like ownership transfer and I thought it was nicer this way.

File: Source/WebCore/editing/FrameSelection.cpp (hunk 2 above)
Comment (cdumez) [reply Source/WebCore/editing/FrameSelection.cpp]: Will fix.

File: Source/WebKit/WebProcess/Network/webrtc/LibWebRTCProvider.h (hunk 3 above)
Comment (cdumez) [reply Source/WebKit/WebProcess/Network/webrtc/LibWebRTCProvider.h]: Ok, I'll see if I can get rid of this include.

Comment (cdumez) [reply Source/WebKit/WebProcess/Network/webrtc/LibWebRTCProvider.h]: I was able to get rid of the new include by moving the constructor & destructor out of line.
